lib/ring.ex

```diff
@@ -0,0 +1,651 @@
+
+    {:reply, reply, %__MODULE__{state | pending_gcs: pending_gcs}}
+  end
+
+  def handle_call(:get_overrides, _from, %{overrides: overrides} = state) do
+    {:reply, {:ok, overrides}, state}
+  end
+
+  def handle_call(:get_nodes, _from, %{nodes: nodes} = state) do
+    nodes = for {node_name, _} <- nodes, do: node_name
+    {:reply, {:ok, nodes}, state}
+  end
+
+  def handle_call(:get_nodes_with_replicas, _from, %{nodes: nodes} = state) do
+    {:reply, {:ok, nodes}, state}
+  end
+
+  def handle_call(:get_pending_gcs, _from, %__MODULE__{} = state) do
+    {:reply, {:ok, Map.keys(state.pending_gcs)}, state}
+  end
+
+  def handle_call({:remove_nodes, node_names}, _from, %__MODULE__{} = state) do
+    has_unknown_nodes? = Enum.any?(node_names, fn name ->
+      not has_node_with_name?(state.nodes, name)
+    end)
+
+    if has_unknown_nodes? do
+      {:reply, {:error, :node_not_exists}, state}
+    else
+      nodes = Enum.reject(state.nodes, fn {name, _} -> name in node_names end)
+      {:reply, {:ok, nodes}, update_nodes(state, nodes)}
+    end
+  end
+
+  def handle_call({:set_nodes, nodes}, _from, %__MODULE__{} = state) do
+    nodes = Node.normalize(nodes, state.replicas)
+    {:reply, {:ok, nodes}, update_nodes(state, nodes)}
+  end
+
+  def handle_call({:set_overrides, overrides}, _from, state) do
+    {:reply, {:ok, overrides}, update_overrides(state, overrides)}
+  end
+
+  def handle_info({:gc, generation}, %__MODULE__{} = state) do
+    pending_gcs =
+      case Map.pop(state.pending_gcs, generation) do
+        {nil, pending_gcs} ->
+          pending_gcs
+
+        {_stale_timer_ref, pending_gcs} ->
+          :ok = do_gc(state.table, generation)
+          pending_gcs
+      end
+
+    {:noreply, %__MODULE__{state | pending_gcs: pending_gcs}}
+  end
+
+  ## Private
+
+  @spec do_find_historical_nodes(
+    key(),
+    hash :: Hash.t(),
+    num :: non_neg_integer(),
+    back :: non_neg_integer(),
+    config :: Config.config()
+  ) :: {:ok, [Node.name()]} | {:error, atom()}
+  defp do_find_historical_nodes(key, hash, num, back, config) do
+    {table, _depth, sizes, generation, overrides} = config
+
+    case Enum.at(sizes, back) do
+      nil ->
+        {:ok, []}
+
+      size ->
+        do_find_nodes_in_table(key, hash, table, overrides, generation - back, size, num)
+    end
+  end
+
+  @spec do_find_nodes(
+    table :: :ets.tid(),
+    generation :: Config.generation(),
+    size :: size(),
+    remaining :: non_neg_integer(),
+    hash :: Hash.t(),
+    found :: [Node.name()],
+    found_length :: non_neg_integer()
+  ) :: [Node.name()]
+  defp do_find_nodes(_table, _generation, _size, 0, _hash, found, _found_length) do
+    # Remaining is now 0, all the requested nodes have been found
+    found
+  end
+
+  defp do_find_nodes(_table, _generation, size, _remaining, _hash, found, size) do
+    # Number of found nodes and number of nodes in the ring are equal, further processing will yield no additional results
```

Comment:
Thank you for the comment. I had to do a double take, but this is so much easier to notice.